Add RFC 1005: Realm key rotation (#5303)
Scille/parsec-cloud: docs/rfcs/1005-realm-key-rotation.md

<!-- Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS -->

# No-reencryption Realm key rotation

## Abstract

This RFC proposes a way to rotate realm keys without requiring re-encryption of all data.

## Background

> The currently realm key rotation system works by replacing the realm key, then re-encrypting

Author: :+1: done.

all realm's vlobs with the new key.

The rational for this approach was that the client only have to deal with a single key at any given time.
However this has some downsides:

> - Reencrypting all vlobs takes some time (though not that much in pratice given it's only

Author: :+1: done.

Now what that means is the client has to in fact deal with up to *two* keys:

1. when a maintenance starts, the client receives the new key
2. it overwrites the realm key with the new one it the user manifest
3. now it tries to access some realm data
   - the maintenance is done, so the new key works. All is good.
   - the maintenance is still ongoing, so the client has to look into the previous
     versions of the user manifest to find the old key. This is messy :'(

## General approach

The idea is as follow:

- remove the whole maintenance logic: a realm is always accessible in read&write
- remove the whole re-encryption logic: realm vlobs are never reencrypted
- key rotation consist of broadcasting the new key in a message to each realm participant
- key rotation also involve the creation of a certificate, this simplifies message
  validation and allow auditability (know when and by who the key rotation occured)
- each key for a given realm has an index (enforced by the server to avoid
  concurrent creation issues)
- vlob create/update/read API commands contains key index information
- user manifest store all keys ordered by index for each realm

> - sharing a realm consist of sending a message containing all keys ordered by index

Reviewer: Does the only up-most recent one is used for decryption/encryption (read&write) whereas the other ones are used only for decryption (read) ?

Author: > Does the only up-most recent one is used for decryption/encryption (read&write) whereas the other ones are used only for decryption (read) ?

yup ;-)

> If that the case, then the old keys will become garbage data (not used for encryption and all the current data use most recent keys).

This only occurs once all the old data have been replaced, which is unlikely: a file is often worked on for some time, then become effectively read-only (i.e. the document is finished and now stays there as reference). Think for instance of the `LICENSE` in a Github repo.

On top of that, old keys are also needed to access the history.

>  Should we consider system to remove useless keys ? (I currently think that it may cause a problem with clients that have a high-rate of renewal of users after some time)

A key is 32bytes long, so a 1mo keys-bundle would contain ~32000 keys, enough to do 10 keys rotation per day during 10 years.
On top of that, we can do un-sharing can be done in batch to only do a single key-rotation.

I would say the issue is more on the data size end: 1mo key bundle and 10 keys rotation per year means 3.5Go of keys-bundle data at the end of the year ! But 1) that some extreme values and 2) it is possible for the server to remove old keys-bundles (old keys-bundles are only used to do self-heal if the last key bundle is corrupted, so it is enough to only keep e.g. the last month of keys-bundles)

Client-side the detection of a new certificate triggers automatically a key rotation
when needed (i.e. user revoked or user role removed).
This is totally transparent for the end user \o/

The elephant in the room is of course compatibility with Parsec v2 (and APIv3),
we go the simple solution here:

- drop compatibility with APIv3 in Parsec v3. A Parsec v2 client is not able to connect
  to a Parsec v3 server and vice-versa.
- keep compatibility with old data. Parsec v2 (both client and server) can be upgraded
  to Parsec v3 without data loss.

## Protocol

### Existing commands changes

The following commands are removed:

- `realm_status`
- `realm_start_reencryption_maintenance`
- `realm_finish_reencryption_maintenance`
- `realm_maintenance_get_reencryption_batch`
- `realm_maintenance_save_reencryption_batch`

> `vlob_poll_changes` and `vlob_list_versions` commands:

Reviewer: ```suggestion
In `vlob_poll_changes` and `vlob_list_versions` commands:
```

Author: done.

- removal of the `in_maintenance` response status

> `realm_update_roles`

Reviewer: ```suggestion
In `realm_update_roles` command:
```

Author: done.

- removal of the `in_maintenance` response status
- add of a `last_key_index` field (with type `Index`). The server must check
  this value is consistent, otherwise the recipient will be missing some keys !
- add of a `bad_last_key_index` reponse status
- `recipient_message` field now contains a `realm_shared_message` instead of a
  `sharing.granted`/`sharing.revoked` one. In case of a role removal this field
  is left empty.

> `vlob_create` & `vlob_update` commands:

Reviewer: ```suggestion
In `vlob_create` & `vlob_update` commands:
```

Author: done.

- removal of the `encryption_revision` field in the request
- add of a `key_index` field (with type `Index`) in the request
- removal of the `bad_encryption_revision` response status
- removal of the `in_maintenance` response status
- add of a `bad_key_index` response status. This error is returend if the key
  index doesn't exist for this realm, or is no longer the last one

> `vlob_read` command:

Reviewer: ```suggestion
In `vlob_read` command:
```

Author: done.

- removal of the `encryption_revision` field in the request
- add of a `key_index` field (with type `Index`) in the ok response
- removal of the `bad_encryption_revision` response status
- removal of the `in_maintenance` response status

> `events_listen` command:

Reviewer: ```suggestion
In `events_listen` command:
```

Author: done.

  initial key is lazily created by doing the very first key rotation (typically during
  minimal sync right after having created the realm).
- key rotation also involves the creation of a certificate, this simplifies message
  validation and allow auditability (know when and by whom the key rotation occurred)
- key rotation consists of creating a new `RealmKeysBundle` with one additional
  key, then submit it (along with the `RealmKeyRotationCertificate` and the per-user
  `RealmKeysBundleAccess`) to the server with the new `workspace_rotate_keys` API command.
- a key rotation may not add a new key: it may also be initiated to replace a previous
  key rotation that was corrupted.
- vlob create/update/read API commands contains a field about the key being used
- user manifest no longer store the realm key, only the realm id and name. The realm keys
  are only fetched when needed (i.e. in case of out/in sync) by using the new
  `workspace_get_keys_bundle` API command.
- `workspace_get_keys_bundle` allows to get previous keys bundles if any, this is needed
  to self-heal in case the new keys bundle is corrupted.

> > TODO: should we use the hash of the key instead of it index ? or an UUID ?

Author: I've removed this part

Client-side, the detection of a new certificate automatically triggers a key rotation
when needed (i.e. user revoked or user role removed).
This is totally transparent for the end user \o/

The elephant in the room is of course compatibility with Parsec v2 (and APIv3). Here we
go the simple way:

- Parsec v2 server only exposes APIv3 (i.e. a Parsec v3 client cannot connect)
- a migration step has to be made on the server database when switching to Parsec v3 server
- once the migration done, only Parsec v3 server can run
- Parsec v3 server only exposes APIv4 (i.e. a Parsec v2 client cannot connect)

> An alternative compatibility mode is [discussed in chapter 7](#73-possible-compatibility-strategy)

Author: :+1:

    "type": "realm_keys_manifest",
    "other_fields": [
        {
            "name": "realm_id",
            "type": "VlobID"
        },
        {
            "name": "keys",
            // Ordered by index
            "type": "List<SecretKey>"
        }
    ]
}
```

TODO: keys bundle system:

- on key-rotation, owner upload a new vlob containing all the decryption keys (including
  the new one), and itself encrypted by a symmetric key. This is the keys bundle.
- the symmetric key is then send by message to all the members of the workspace.
- each member update it user manifest with the vlob id and key of the keys bundle.

Given the keys bundle is stored as a vlob, anyone with write access can upload new versions
of it, so the client always only retrieve the version 1 of the vlob.

Three benefits for this approach:

- Keys are encrypted a single time instead of for each member: faster to do and less
  disk space consumed (it must be stored for each user as message and in user manifest)
- When access is removed, the user is no longer able to get back the decryption keys
  (instead of having them in it user manifest). The client can of course save them when

>   it has access to them, but it requires a modified Parsec.

Reviewer: In that case the user would only be able to read the data that where using the previous decryption keys ? Without a modified parsec, the user could still copy the data on another support.

Author: > In that case the user would only be able to read the data that where using the previous decryption keys ?

Yes, no matter what new data are not accessible to him

> Without a modified parsec, the user could still copy the data on another support.

Yes totally, hence no having the key stored on the local machine is only a hardening element

- The decryption keys are no longer at rest on the client: they are only downloaded

>   when the client needs them (i.e. when in/out sync is needed) and kept in memory

Author: all data stored in local are encrypted by a different key (the local_symkey stored)

>   when the client needs them (i.e. when in/out sync is needed) and kept in memory

Reviewer: That mean the user cannot access its data offline anymore ? that was a nice feature :thinking:

>   when the client needs them (i.e. when in/out sync is needed) and kept in memory

Author: > That mean the user cannot access its data offline anymore ? 

No it's still possible: all data stored in local are encrypted with a dedicated key (which is stored in the device keys file)

>   when the client needs them (i.e. when in/out sync is needed) and kept in memory

Reviewer: :+1:

client must make sure the `key_index` field strictly follow the value of the
previous key rotation certificate for this realm. (i.e. we reject the
certificate if there is a gap in key indexes)

## 5 Changes: Server-side

### 5.1 Change HTTP `Server` header

Currently the HTTP `Server` value is `parsec`, we should change it to `parsec/3` to
simplify the detection of an incompatible server (see
[7.1 - No Parsec v2 to v3 auto-update](#71-no-parsec-v2-to-v3-auto-update) ).

### 5.2 Datamodel

The main issue with data migration is handling realm currently under maintenance given
it asks the question of which key to keep:

- Keeping only the last key doesn't work as some data are not encrypted for it
- Keeping the n-1 keep only makes migration complicated on client-side (the server
  no longer has concept of realm under migration, however the client has to find
  a way to know that. On top of that the migration process requires accessing
  older version of the user manifest which add complexity)
- Keeping both n-1 and n adds basically the same issues with client-side

So we go simple on this and only allow server migration if all the realms have

> finished they maintenance, or if a realm doesn't contain any data (see below).

Reviewer: ```suggestion
finished their maintenance, or if a realm doesn't contain any data (see below).
```

Author: :+1: done.

To avoid being blocked by a workspace forever in maintenance (e.g. workspace owned by
a single user that got revoked while it was re-encrypting it), we also provide a cli for
removing realm vlobs (cannot remove a realm given it is related to certificates) or
organization form the database.

### 5.3 Server data migration

Drop tables:

- `vlob_encryption_revision`
- `message`

Table `vlob_atom`:

- Remove all rows that contains old encryption revision data (i.e. the table only contains
  the vlob currently accessible by the users).
- Remove column `vlob_encryption_revision`
- Add column `key_index` (`INTEGER NOT NULL`). This defaults to 1 when migrating data.

Table `realm`

- Remove columns `encryption_revision`, `maintenance_started_by`, `maintenance_started_on` and `maintenance_type`

## 6 Changes: Client-side

### 6.1 Certificate Ops

`RealmKeyRotationCertificate` should be checked (only OWNER of the realm can create such

> certificate).

Author: > Should we include the admin ? since they can remove / add user.
> If we remove an user we should encrypt the new data using a new key ASAP.

This is handled by a new "key rotation manager " that relies on the certificate ops to detect revoked user and then remove they access and do a key rotation.

> That would be cumbersome if when we remove an user, we need to wait for the owner to connect to generate a new set of keys

We have no choice on this: generating a new key can only be done by a user with access to the realm, but ADMIN profile is orthogonal to that.

### 6.2 Fetching keys bundle

The client lazily fetch the realm keys bundle: it do the fetch operation when a server

> operation requires it (i.e. `realm_share`, `vlob_read`, `vlob_create` and `Vlob_update`).

Author: :+1: done.

The realm keys bundle is only kept in RAM, not on disk.

When fetching keys bundle and keys bundle access, both must be checked against the
corresponding `RealmKeyRotationCertificate` (same author and timestamp).

At any time, `realm_share`/`vlob_read`/`vlob_create` can return `bad_key_index` if the client
consider the wrong key to be the last one. In this case the client simply has to forget
it realm keys bundle and re-do the fetch operation.

In case the current keys bundle is corrupted (e.g. bad signature, missing keys etc.),
a self-healing is attempted:

1. client looks into the certificates to find previous key rotation certificates
2. client calls `realm_get_keys_bundle` with a `timestamp` parameter set to the key
   rotation certificate.

>    - If keys bundle is corrupted, go to 1. until there is no more keys bundles

Reviewer: ```suggestion
   - If keys bundle is corrupted, go to step 1. until there is no more keys bundles
```

Author: :+1: done.

   - If keys bundle can be retrieve, the self-healing has succeeded.
   - If the keys bundle was never available to our user (i.e. wasn't the last one when
     the realm got shared to us), the self-heading has failed.

3. Client dispatch internally an event to inform of corruption (including the author
   of the faulty key rotation) and self-healing outcome. This event typically ends
   up as a warning logged and gets displayed in the GUI.

### 6.3 Datamodel

> TODO

### 6.4 Automatic key rotation

For automatic key rotation, a new "key rotation manager" is to be introduced. It role
is very simple:

1. Wait for new `UserRevoked`, `RealmRoleUpdate` and `RealmKeyRotationCertificate` certificates
2. For `UserRevoked`, call `realm_unshare` for each realm it is owner (and the revoked
  user is part of)
3. For `RealmRoleUpdate`, if the role is None and the user is owner of the realm:

>   3.1. Wait a random time between 30 and 60 seconds (and reset the wait if we were
>        already waiting for this realm)
>   3.2. Call `realm_rotate_key`

Reviewer: ```suggestions
  a. Wait a random time between 30 and 60 seconds (and reset the wait if we were
       already waiting for this realm)
  b. Call `realm_rotate_key`
```

Author: I think it's better to keep number instead of letter here given 3.1 and 3.2 are ordered (and not two separated possibilities)

Reviewer: We can keep the number, it's more about the numbering format used that isn't supported by markdown (it wont be presented as a sub-list)

```suggestion
  1. Wait a random time between 30 and 60 seconds (and reset the wait if we were
       already waiting for this realm)
  2. Call `realm_rotate_key`
``` 

or 

```suggestion
  1) Wait a random time between 30 and 60 seconds (and reset the wait if we were
       already waiting for this realm)
  2) Call `realm_rotate_key`
```

> 4. For `RealmKeyRotationCertificate` cancel the wait from step 3.1 if any

Reviewer: ```suggestion
4. For `RealmKeyRotationCertificate` cancel the wait from step 3.a if any
```

Author: see above

> The wait at step 3.1 is here for two reasons:

Reviewer: ```suggestion
The wait at step 3.a is here for two reasons:
```

Author: see above

After that it will deny any APIv3 request, hence forcing all it users to upgrade
they Parsec client at once.

### 7.1 No Parsec v2 to v3 auto-update

Given switching from Parsec v2 to v3 breaks the way ones connect to the server, they
are to be considered as different products:

- On Linux, the `parsec` [Snap allows parallel install](https://forum.snapcraft.io/t/parallel-installs/7679).
- On Windows/Mac, Parsec v2 shouldn't prompt "a new version is available" about Parsec v3.
- On Windows/Mac, side-by-side installation of Parsec v2 and v3 should be possible.
- A new version of Parsec v2 should be released that detects when we try to connect to
  a Parsec v3 server (i.e. request to `/ws` lead to a 404 and `Server` header is `parsec/3`)
  and prompt a message about installing Parsec v3 client (explaining the Snap parallel
  trick if on Linux).

In case of parallel install, the automatic handling of `parsec://` links may break.

> TODO: What should we do ?
>
> 1) don't care (might be already handled by the OS)
> 2) introduce `parsec3://` protocol ?

### 7.2 Server-side organization migration

This is typically for the SAAS version Parsec: some clients may want to be able

> to contrôle when the switch from v2 to v3 should come for they organization.

Reviewer: ```suggestion
to control when the switch from v2 to v3 should come for they organization.
```

Author: :+1: done.

<!-- Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS -->

# No-reencryption Realm key rotation

## 0 Abstract

This RFC proposes a way to rotate realm keys without requiring re-encryption of all data.

## 1 Background

The current realm key rotation system works by replacing the realm key, then re-encrypting
all realm's vlobs with the new key.

The rational for this approach was that the client only have to deal with a single key at any given time.
However this has some downsides:

- Reencrypting all vlobs takes some time (though not that much in practice given it's only
  metadata, the empirical rule being 1000 times less metadata than actual data in Parsec).
- This concept must be exposed to the user so that he can decide when to do the reencryption
  maintenance procedure. This is a concept complex to explain so not great for UX :/

> - A realm encryption revision index is kept server-side and send to the client to ensure
>   it is not using the wrong key to decrypt

Reviewer 2: Not sure I understand this. Is this index related to the fact that, because of maintenance, there may be two keys? 
If so, then it is already explained below (the last paragraph of section 1) and I would simply remove the whole item from the list here.

```suggestion
```

- A re-encryption API is needed (the `realm_(start|stop)_reencryption_maintenance` &
  `vlob_maintenance_(get|save)_reencryption_batch` commands)
- The concept of maintenance has been introduced: a realm under maintenance cannot be read/write
- Due to the maintenance, a reencryption must finish no matter what. So all realm owner have to

>   watch for ongoing maintenance an participate in it (even if they are not the initiator of it),

Reviewer 3: `and` participates

Reviewer 3: ```suggestion
  watch for ongoing maintenance and participate in it (even if they are not the initiator of it),
```

>   watch for ongoing maintenance an participate in it (even if they are not the initiator of it),

Reviewer 2: The last two items in the list should be merged (they both refer to the maintenance).

Also some typos : "watch for ongoing maintenance an**d** participate~s~"

>   just in case the initiator disconnects (to be never seem again) before completion.

Reviewer 3: ```suggestion
  just in case the initiator disconnects (to be never seen again) before completion.
```

On top of that the added security is small (remember, this was a practical choice) given
the client could have downloaded all metadata when it has access (typically when doing
file content a search on the workspace mountpoint).
In addition since Parsec 2.16 the manifests are eagerly downloaded (instead of the
previous lazy approach), so the reencryption really has no security benefit.

And now the final blow: the maintenance preventing read/write access was such a strong
limitation that we ended up adding a feature to keep read access possible.
Now what that means is the client has to in fact deal with up to *two* keys:

1. when a maintenance starts, the client receives the new key

> 2. it overwrites the realm key with the new one it the user manifest

Reviewer: Propose suggestion ;)

> 2. it overwrites the realm key with the new one it the user manifest

Reviewer 3: ```suggestion
2. it overwrites the realm key with the new one in the user manifest
```

3. now it tries to access some realm data
   - the maintenance is done, so the new key works. All is good.
   - the maintenance is still ongoing, so the client has to look into the previous
     versions of the user manifest to find the old key. This is messy :'(

## 2 General approach

The idea is as follow:

- remove the whole maintenance logic: a realm is always accessible in read&write
- remove the whole re-encryption logic: realm vlobs are never reencrypted

> - remove the whole message logic: we use a special ad-hoc command to access the

Reviewer 2: The "message logic" should be briefly introduced/explained in section 1. I'll continue reading to understand ^^

> - remove the whole message logic: we use a special ad-hoc command to access the

Author: ```suggestion
- remove the whole message logic (i.e. sending arbitrary message between users, currently only used to communicate workspace keys during sharing): we use a special ad-hoc command to access the
```

  realm's keys (see below)
- a realm now has it manifests encrypted with multiple keys, at any time a new
  key can be added

> - each key for a given realm has an index

Reviewer: ```suggestion
- each keys for a given realm has an index
```

Author: see https://github.com/Scille/parsec-cloud/pull/5303#discussion_r1355209180

> - only the last key should be used to encrypt new manifest

Reviewer: ```suggestion
- only the last key should be used to encrypt new manifests.
```

- all those keys are serialized together in a `RealmKeysBundle` which is encrypted by the
  "keys-bundle key" symmetric key and stored in a dedicate table in the server database
- the "keys-bundle key" is serialized in a `RealmKeysBundleAccess`. This document is signed

>   by the author of the key rotation and encrypted for each member of the realm with their

Reviewer: ```suggestion
  by the author of the key rotation and encrypted for each members of the realm with their
```

Reviewer 2: singular (member) is fine in this case

  public key
- for sharing a realm, the last `RealmKeysBundleAccess` is encrypted for the new user.

>   The `RealmKeysBundleAccess` also contains the workspace label.

Author: workspace name is needed to be display something readable to the user in the GUI, but it is a sensible information so it cannot be unencrypted like the realm id is

- keys-bundle & keys-bundle access are never stored on-disk client-side: they are instead
  lazily fetched from server when needed and kept on RAM.
- realm initial key is lazily created by doing the very first key rotation (typically during
  minimal sync right after having created the realm).
- key rotation also involves the creation of a `RealmKeyRotationCertificate` certificate to
  allow auditability (knowing when and by whom the key rotation occurred)
- key rotation certificate contains the index of the key along with it algorithm and a

>   canary (i.e. an encrypted payload to ensure the key obtained from the keys-bundle is genuin)

Reviewer 2: ```suggestion
  canary (i.e. an encrypted payload to ensure the key obtained from the keys-bundle is genuine)
```

- key rotation consists of creating a new `RealmKeysBundle` with one additional
  key, then submit it (along with the `RealmKeyRotationCertificate` and the per-user
  `RealmKeysBundleAccess`) to the server with the new `workspace_rotate_keys` API command.

> - given each key rotation is expected to add a single key to the bundle, the key index is

Reviewer 2: singular is OK here

>   also used to identify the key rotation and the keys-bundle

Reviewer 2: This smells a little bit... using the index to identify key rotation and keys-bundle imposes a strong constraint. Are we sure there can never be "holes"? I don't have a counter-example yet, but I'm thinking in some data corruption or alike...

What about using hashes for identification + tracking the last hash (rotation)? They can even be chained by having pointers to the "parent" rotation (yes, a git-ish solution ^^). It is certainly more complex, but it might be worth it to at least compare it to the index-based solution.

Author: > What about using hashes for identification

using a hash of a key is potentially worrying given it can leak info if not done properly

hence the only other solution is using a uuid to identify a key, but given we want key and key-rotation-certificate to be related 1-to-1 (for simplification), then it's more logical to identify them by the order they are created

Author: > Can a KeyBundle be added/created for another reason besides a key rotation? (Maintenance, Security, Data suppression...)

for simplification we don't want that !

> If we consider data suppression, there might be some cases where keys are no longer needed

In theory yes, in practice it's most likely there is still some data that use any given key (think of the README in a repo that haven't been updated in ages, also in Parsec we handle History so old data must still be accessible).
In the end it's no use bothering with cleanup up old keys.

- vlob create/update/read API commands contains a field with the index of the key being used

> - user manifest no longer store the realm key, only the realm id and name. The realm keys

Reviewer 3: `stores`

Reviewer 3: ```suggestion
- user manifest no longer stores the realm key, only the realm id and name. The realm keys
```

  are only fetched when needed (i.e. in case of out/in sync) by using the new
  `workspace_get_keys_bundle` API command.
- `workspace_get_keys_bundle` allows to get previous keys-bundles if any, this is needed
  to self-heal in case the new keys-bundle is corrupted.
- in case the "keys-bundle key" becomes corrupted, the self-heal operation consist of

>   retrieving the last valid "keys-bundle key" and doing a new key rotation from there.

Author: > Unless the server is capable of detected the problem itself 

`KeyBundle` is encrypted so server has not power here !

> the last key bundle is requested and, IF a problem arises, the previous key bundles are requested and a new key rotation is performed
>
> Should we consider "tagging" the corrupted key bundle so it is no longer used and/or can be cleaned later?

You are correct: a corrupted key rotation could introduce a key that doesn't correspond to the key rotation certiifcate's canary. In this case the self-heal operation must include the key (given it index is the position in the keys array), but doesn't know what to include !

The solution is to sign the keybundle (so that we trust whatever key are in the key bundle, even if some are missing), and allow missing keys (providing the keys in the `KeyBundle` as `Vec<Option<Bytes>>`

>   retrieving the last valid "keys-bundle key" and doing a new key rotation from there.

Author: > The solution is to sign the keybundle (so that we trust whatever key are in the key bundle, even if some are missing), and allow missing keys (providing the keys in the KeyBundle as Vec<Option<Bytes>>

I've slightly changed the solution: It is much simpler to just consider a KeyBundle valid as long as it is correctly signed and contains the right number of keys. This way there is no need to handle a `None` value for the keys, and the self-heal operation becomes trivial: just recursively get the previous key bundle until one of them is valid.

![Architecture schema](./1005-realm-key-rotation/schema.svg)

1) put it in the `KeysBundleAccess`: close to the current solution, but seems like a hack:
   a key rotation is sometime done only to do something totally different than key rotation,
   should this field be ignored once we already have the workspace name set in the client ?
2) put it in the `KeysBundle`: allow to have an owner globally renaming the workspace but
   still a hack, need to fetch the keys-bundle on every key rotation to ensure the name
   hasn't changed ?
3) put it in a dedicated certificate: allow clean history the renaming but more complex
   and need to encrypt the info within the certificate

> 4) add a dedicate field in the `realm_share` command to provide it

Author: yes this should we rewritten (especially given we settled on the dedicated command approach)

Client-side, the detection of a new certificate automatically triggers a key rotation

> when needed (i.e. user revoked or user role removed).

Author: >  Or the automatic triggers has additional constraint that may not trigger a key rotation 🤔

Yes, the client can detect new certificate which doesn't require key rotation (e.g. a user create certificate)

> when needed (i.e. user revoked or user role removed).

Reviewer 2: Not sure about the "detection" part (or maybe I misunderstood):
- Who is actually "triggering" the key rotation? I suppose is the client performing the action that leads to a new certificate (e.g. revoke of a user)
- In that case, is not really on client-side "detection" (other clients will detect the certificate but will not trigger the key rotation)

This is totally transparent for the end user \o/

The elephant in the room is of course compatibility with Parsec v2 (and APIv3). Here we
go the simple way:

- Parsec v2 server only exposes APIv3 (i.e. a Parsec v3 client cannot connect)
- a migration step has to be made on the server database when switching to Parsec v3 server

> - once the migration done, only Parsec v3 server can run

Reviewer 3: the migration `is` done

Reviewer 3: ```suggestion
- once the migration is done, only Parsec v3 server can run
```

- Parsec v3 server only exposes APIv4 (i.e. a Parsec v2 client cannot connect)

An alternative compatibility mode is [discussed in chapter 7.2](#72-possible-compatibility-strategy)
This is however complex to implement and test, so shouldn't be implemented it unless

> explicitly needed.

Author: ¯|_(ツ)_/¯​​

>   single key rotation can be issued

### 3.4 New command: `realm_rotate_key`

```json5
[
    {
        "major_versions": [4],
        "req": {
            "cmd": "realm_rotate_key",
            "fields": [
                {
                    // Signed `RealmKeyRotationCertificate` certificate,
                    // contains realm_id & key_index
                    "name": "certificate",
                    "type": "Bytes"
                },
                {
                    // Per-participant (including self) signed&encrypted
                    // `RealmKeysBundleAccess` document, contains the key to
                    // decrypt `keys_bundle` field
                    "name": "per_participant_keys_bundle_access",
                    "type": "Map<UserID, Bytes>"
                },
                {
                    // Signed&encrypted `RealmKeysBundle` document, contains all the
                    // keys ever in use for this realm
                    "name": "keys_bundle",
                    "type": "Bytes"
                }
            ]
        },
        "reps": [
            {
                "status": "ok"
            },
            {

>               // If the user doesn't have OWNER role in the realm
>                 "status": "not_allowed"

Reviewer: ```suggestion
                // If the user doesn't have OWNER role in the realm
                "status": "not_allowed"
```

                    },
                    {
                        "name": "ballpark_client_late_offset",
                        "type": "Float"
                    },
                    {
                        "name": "backend_timestamp",
                        "type": "DateTime"
                    },
                    {
                        "name": "client_timestamp",
                        "type": "DateTime"
                    }
                ]
            }
        ]
    }
]
```

> Notes:
>
> - Each `realm_rotate_key` command is expected to provide a `RealmKeysBundle` that differs
>   from the previous one by exactly appending a single key. This will be checked client-side.
> - Certificate, keys-bundle and keys-bundle access must all be signed by the same author
>   with the same timestamp. This will be checked client-side.
> - Only `realm_rotate_key` command updates the keys-bundle, hence a newly created realm
>   has no keys-bundle.

### 3.5 New command: `realm_get_keys_bundle`

```json5
[
    {
        "major_versions": [4],
        "req": {
            "cmd": "realm_get_keys_bundle",
            "fields": [
                {

>                     "name": "key_index",

Author: missing realm_id field

                    },
                    {
                        // Signed&encrypted `RealmKeysBundle`
                        "name": "keys_bundle",
                        "type": "Bytes"
                    }
                ]
            },
            {
                // User has no access on this realm
                "status": "not_allowed"
            },
            {
                // No keys-bundle set so far (i.e `realm_rotate_keys` has never been called)
                "status": "no_keys_bundle"
            },
            {
                // `certificate_index` argument doesn't correspond to a `RealmRotateKeyCertificate`
                "status": "not_found"
            }
        ]
    }
]
```

As stated before, client is expected to call `realm_get_keys_bundle` the first time it
needs to fetch/upload vlob to/from the server, and to store the resulting keys in RAM.

This brings three benefits:

- Keys are no longer stored in each user manifest, which save storage space and avoid
  having to sync a new version of the user manifest each time a key rotation is done.
- When access is removed, the user is no longer able to get back the decryption keys

>   (instead of having them in it user manifest). The client can of course save them when

Reviewer: ```suggestion
  (instead of having them in its user manifest). The client can of course save them when
```

  it has access to them, but it requires a modified Parsec client.
- The decryption keys are no longer at rest on the client storage.

## 4 Changes: Data

### 4.1 Existing data: messages

All the messages types are removed:

- `sharing.granted`
- `sharing.reencrypted`
- `sharing.revoked`

On top of that, a special logic is added in the client to ignore those messages
types.

> **Note:**
>
> Ignoring those messages means the client may end up with invalid data (e.g. a

> > user hasn't started it client in a while, he does it after the Parsec v3 migration

Reviewer: ```suggestion
> user hasn't started its client in a while, he does it after the Parsec v3 migration
```

> and now all the workspaces it has been part of that have been re-encrypted in the
> meantime fail to load with an "decryption error" error message...).
>

> > However those error can simply be fixed by removing and re-adding the access rights

Reviewer 3: those `errors`

Reviewer 3: ```suggestion
> However those errors can simply be fixed by removing and re-adding the access rights
```

> to the user, which is the obvious "IT, have you tried to turn it off and on again ?" way.

### 4.2 Existing data: `UserManifest`

Remove `last_processed_message` field.
Modify `workspaces` field, i.e. the `WorkspaceEntry` nested type:

```json5
{
    "label": "WorkspaceEntry",
    "fields": [
        {
            "name": "id",
            "type": "VlobID"
        },
        {
            "name": "name",
            "type": "EntryName"
        }
        // Removed fields:
        // - key
        // - encryption_revision
        // - encrypted_on
        // - role_cached_on
        // - role_cache_value
    ]
}
```

> **Note:**
>
> `role_cached_on` and `role_cache_value` are not directly related to this RFC,
> but given they are no longer in use in Parsec v3, we also remove them.

> **TODO:**
>
> Should we rename workspace name into label ?
>
> We have "Human handle label" and "Device label", the "label" refers to the
> fact they are purely display information (i.e. not used for unique identification).
>
> So given the idea is similar here (this field can even be modified by the user !),

> > it makes sens to rename it to "workspace label".

Reviewer 3: `sense`

Reviewer 3: ```suggestion
> it makes sense to rename it to "workspace label".
```

>
> On the other hand keeping "workspace name" is less work, and the fact this name
> can be modified where the human handle / device label cannot may justify this

> > is a different concept...

Reviewer: TODO

Author: "workspace name" works differently than human handle label & device label:
- human handle / device label cannot be updated
- human handle / device label are shared among all organization users

hence the renaming would bring confusion, so it's a no !

        },
        {
            "name": "realm_id",
            "type": "VlobID"
        },
        {
            "name": "key_index",
            "type": "Index"
        },
        {
            "name": "encryption_algorithm",
            "type": "SecretKeyAlgorithm"
        },
        {
            "name": "hash_algorithm",
            "type": "HashAlgorithm"
        },
        {
            // Empty message encrypted with the key, this is used to ensure the
            // the key retrieved in the keys-bundle hasn't been tempered (given
            // each key rotation author generates a new keys-bundle from the
            // previous bundle and his one additional key).
            //
            // Note since the message is empty, we are doing here AEAD-without-data
            // given we only care of the MAC part.
            "name": "key_canary",
            "type": "Bytes"
        }
    ]
}
```

When processing this certificate (in addition to regular access checks), the
client must make sure the `key_index` field strictly follow the value of the
previous key rotation certificate for this realm. (i.e. we reject the
certificate if there is a gap in key indexes)

> `SecretKeyAlgorithm` has one possible values: `XSALSA20-POLY1305`.

Reviewer: Since this is new stuff we could rely on algorithm that has been audited in `RustCrypto` ?

That way we could plan to remove `sodiumoxide` in the dependency of `libparsec_crypto` that would simplify its implementation, dependencies specification, tests and packaging in the long run

Author: > Since this is new stuff we could rely on algorithm that has been audited in RustCrypto ?

Yes that's the idea: switch to ietf's chacha20-poly1305 which is much more scrutinized as it is used for TLS (for instance RustCrypto implementation has already been audited)

`HashAlgorithm` has one possible value: `SHA256`.

> **TODO:**
>

> > The fields `encryption_algorithm` & `hash_algorithm` allows for future evolution.

Author: A new RFC will be opened to discuss this

The idea is basically to modify file manifest to references the file blocks as `(block_id, key_index, block_hash)`, this way all data using symmetric encryption will use the same key

Reviewer: Did you create an issue to trace the requirement of the new RFC ?

> > The idea would be to switch to  Chacha20-Poly1305 given it is a more scrutinized

Reviewer 3: remove extra space

Reviewer 3: ```suggestion
> The idea would be to switch to Chacha20-Poly1305 given it is a more scrutinized
```

> algorithm with audited implementations.
>

> > Another benefit is to be able to easily upgrade those algorithm in the future in

Reviewer 3: those `algorithms`

Reviewer 3: ```suggestion
> Another benefit is to be able to easily upgrade those algorithms in the future in
```

> > case of exploit in them.

Reviewer 3: `of` them ?

Reviewer 3: algorithms in the future in case of exploitation of them.

Author: `algorithm` -> `algorithms`

Author: ` in case of exploit in them.` -> ` in case of an exploit in them.`

        {
            "name": "realm",
            "type": "VlobID"
        },
        {
            "name": "keys",
            // Keys ordered by index
            "type": "List<Bytes>"
        }
    ]
}
```

### 4.5 New document `RealmKeysBundleAccess`

```json5
{
    "label": "RealmKeysBundleAccess",
    "type": "realm_keys_bundle_access",
    "other_fields": [
        {
            "name": "author",
            "type": "DeviceID"
        },
        {
            // Timestamp must be equal to the one in the `RealmRoleCertificate`
            // or `RealmKeyRotationCertificate` (depending if the access has been
            // given before or after last key rotation)
            "name": "timestamp",
            "type": "DateTime"
        },
        {
            "name": "realm",
            "type": "VlobID"
        },
        {
            "name": "key_index",
            "type": "IndexInt"
        },
        {
            "name": "keys_bundle_key",
            "type": "SecretKey"
        },
        {
            // Only present when the `RealmKeysBundleAccess` is generated for use with
            // `realm_share` command (and not `realm_rotate_key` given in this case the

>             // user already )

Reviewer 3: already what ?

Author: ![image](https://github.com/Scille/parsec-cloud/assets/3187637/06c632e2-fa15-4e46-a2a0-7362c20abbe3)

Author: this is outdated stuff and will be remove now that we settled on the go light approach for `KeyBundleAccess`

            "name": "workspace_label",
            "type": "RequiredOption<String>"
        }
    ]
}
```

> **TODO:**
>
> Do we need author/timestamp/realm_id in `KeysBundle`&`KeysBundleAccess` ?
>
> The need is to ensure they both correspond to a given `RealmKeyRotationCertificate`, but
> adding those fields seems a lot of duplicated data :'(
>
> And we cannot simply add `KeyBundle`&`KeysBundleAccess` hashes to the certificate
> given they contains encryption keys (and we cannot do HMAC).
>
> Another solution is to "go light":
>
> - `KeysBundleAccess` & `KeysBundle` are not signed (only encrypted)
> - `KeysBundleAccess` only contains `keys_bundle_key` and `workspace_label`
> - `KeysBundle` only contains the keys
> - if `KeysBundleAccess` is forged then we cannot decrypt `KeysBundle`
> - if `KeysBundle` is forged, then the key canary in the certificate doesn't work
>

> > Do you see a flaw here ?

Reviewer: TODO

Reviewer: > The need is to ensure they both correspond to a given `RealmKeyRotationCertificate`, but
> adding those fields seems a lot of duplicated data :'(

Certificate are identified with a serial, we could do something similar with an UUID. That would removing the need to specify the fields `author&co` everytime but only a single field (`id|bundle_id|serial|bundle_serial` :shrug:).

Author: > Certificate are identified with a serial

Not sure what you mean, currently there is no serial number on the certificates (they have an index based on their order by timestamp)

> we could do something similar with an UUID. That would removing the need to specify the fields author&co everytime but only a single field (id|bundle_id|serial|bundle_serial 🤷).

The fields author&timestamp are useful to validate a signature. Here we can get those fields back from the related certificate, but it makes things more complex...
Another solution is to only encrypt the keysbundle & keysbundleaccess, but then the bundle_id/bundle_serial values are not guaranteed to be the right one... unless we take into account the encryption layer (given we do AEAD), in this case we are back to the "go light" solution, do you agree ?

Author: we are good for the "go light" approach :+1:

3) No entry in the user manifest, but a realm role certificate (with non-null role)
   exists: corresponds to a newly shared realm

Case 1 and 2 work just fine: the workspace name is present in the user manifest.

Case 3 requires to go fetch the workspace name from the `KeysBundleAccess` that
correspond to the key index when the realm was shared with our user.

> **TODO:**
>
> It's most likely last sentence made you scream internally considering the complexity
> of the operation...
>
> The alternatives are:
>
> - always provide the workspace label in the `KeysBundleAccess` (and not only for
>   `KeysBundleAccess` provided while calling `realm_share`). The downside is this is
>   wasteful in server database storage.
> - encrypt the workspace label with the recipient public key and put it in the realm role
>   certificate. It is also wasteful, both for the server and the client that store every
>   certificate.
> - pass the workspace label as a separated encrypted (with recipient public key) field
>   in `realm_share` command, then provide a `realm_get_workspace_label` command to
>   retrieve it. This add very ad-hoc APIs that will most likely be deprecated when we

> >   will introduce workspace label controlled by workspace owners.

Author: > Why would that depreciate the command realm_get_workspace_label ? if a workspace owner change the name of a workspace we need a command to get the current workspace label, no ?

You are correct, but the system would most likely work differently given right now we only have to deal with a per-user workspace ~label~ ~name~ nick :trollface: that can be modified freely.
In the future we'd like to have the owner managing the actual workspace name (including modifying it) and each user still able to rename the workspace to it liking.

We are not very clear how to implement that and we don't want to think too much about it right now. So if we introduce `realm_get_workspace_label`, we have to agree it is not at all future proof ^^

> >   will introduce workspace label controlled by workspace owners.

Author: Solution selected is the "pass the workspace label as a separated encrypted" with a dedicated `realm_get_workspace_name` command

Case 3 should be handle by the user manager:

> - On init the user manager compare the user manifest with the realm roles certificates

Reviewer 3: realm `role` certificates

Reviewer 3: ```suggestion
- On init, the user manager compares the user manifest with the realm role certificates
```

  to detect realm with missing entry

> - The user manifest also listen for new certificates event to re-do the check

Reviewer 3: `listens`

Reviewer 3: ```suggestion
- The user manifest also listens for new certificates event to re-do the check
```

> - The user manifest then fill the missing entries using `realm_get_keys_bundle` server command

Reviewer 3: `fills`

Reviewer 3: ```suggestion
- The user manifest then fills the missing entries using `realm_get_keys_bundle` server command
```

- When a workspace listing is done, the missing entries are simply ignored ()

> ### 6.3 Fetching keys-bundle

Reviewer: I'm not confident about having to fetch the bundle each time (given it simplify not having to handle a cache for it). 
That require to do the keys check every time we start parsec and fells like `rust-analyzer` cold start when I start my IDE

Reviewer: At first I was thinking that `Fetching keys-bundle` was made at the start of the application, but will likely be done when we need to upload/download some data in that case that would add latency on the user for the first sync :thinking:

Author: The fetch keys bundle is only needed when uploading/downloading data to/from the server.
Hence when we do the fetch keys bundle we were about to do a server request no matter what, so having two request instead of one the first time it no big deal (adding a in-ram cache on this is totally trivial).

The client lazily fetch the realm keys-bundle: it does the fetch operation when a server
operation requires it (i.e. `realm_share`, `vlob_read`, `vlob_create` and `vlob_update`).

The realm keys-bundle is only kept in RAM, not on disk.

> Once fetched, keys-bundle must be checked against it corresponding `RealmKeyRotationCertificate`

Reviewer: ```suggestion
Once fetched, keys-bundle must be checked against its corresponding `RealmKeyRotationCertificate`
```

> for same author/timestamp/realm_id/key_index.

Reviewer: will change with the _TODO: specifying those author, timestamp, realm_id fields every-time_ ?

Author: yes totally

> Each key in the Keys-bundle must be checked against the canary in `RealmKeyRotationCertificate`

Reviewer: ```suggestion
Each keys in the Keys-bundle must be checked against the canary in `RealmKeyRotationCertificate`
```

that introduced it.

> Once fetched, keys-bundle access must be checked against it corresponding

Reviewer: ```suggestion
Once fetched, keys-bundle access must be checked against its corresponding
```

`RealmKeyRotationCertificate` or `RealmUpdateRoleCertificate` (if the role was given
after last key rotation) for same author/timestamp.

At any time, `realm_share`/`vlob_read`/`vlob_create` can return `bad_key_index` if the client

> consider the wrong key to be the last one. In this case the client simply forget it realm

Reviewer 3: `forgets`

Reviewer 3: ```suggestion
consider the wrong key to be the last one. In this case the client simply forgets it realm
```

keys-bundle and re-do the fetch operation.

### 6.3 Keys-bundle self-healing

A keys-bundle can be corrupted in multiple ways:

1. Additional keys
2. Missing or tempered keys
3. Invalid data: bad signature, bad encryption, bad serialization

In case of 1, self-healing is easy: just remove the keys that don't have a corresponding
key rotation certificate.

In case of 2 or 3, the fix goes:

1. client looks into the certificates to find previous key rotation certificates
2. client calls `realm_get_keys_bundle` with a `key_index` parameter set to the key
   rotation certificate.

   - If keys-bundle is corrupted, go to step 1. until there is no more keys-bundles
   - If keys-bundle can be retrieve, the keys in the bundle should be checked against
     the canary in the `RealmKeyRotationCertificate`
   - If the keys-bundle was never available to our user (i.e. wasn't the last one when

>      the realm got shared to us), the self-heading has failed.

Reviewer: ```suggestion
     the realm got shared to us), the self-healing has failed.
```

In any case, client should dispatch internally an event to inform of corruption
(including the author of the faulty key rotation) and self-healing outcome. This event
typically ends up as a warning logged and gets displayed in the GUI.

> **Note:**
>
> - We cannot simply get the keys-bundle that originally created a given key given
>   the realm might have been shared with the user later on.
> - A more powerful algorithm would be to combine the valid keys in each keys-bundles.
>   (e.g. we can self-heal if bundle A contains key X corrupted and bundle B contains
>   key Y corrupted). In practice however it is likely a bundle is either totally
>   valid or totally broken.

### 6.4 Automatic key rotation

> For automatic key rotation, a new "key rotation manager" is to be introduced. It role

Reviewer: ```suggestion
For automatic key rotation, a new "key rotation manager" is to be introduced. Its role
```

is very simple:

1. Wait for new `UserRevokedCertificate`, `RealmRoleCertificate` and
   `RealmKeyRotationCertificate` certificates
2. For `UserRevokedCertificate`, call `realm_unshare` for each realm we are owner (and
   the revoked user is part of)

> 3. For `RealmRoleCertificate`, if the role is None and ware are owner of the realm:

Reviewer: ```suggestion
3. For `RealmRoleCertificate`, if the role is None and we are owner of the realm:
```

  3.1. Wait a random time between 30 and 60 seconds (and reset the wait if we were
       already waiting for this realm)
  3.2. Call `realm_rotate_key`
4. For `RealmKeyRotationCertificate` cancel the wait from step 3.1 if any

The wait at step 3.1 is here for two reasons:

- Give priority to the author of the certificates to do the key rotation. This is
  typically needed if he is doing un-sharing batch (and hence will do the key rotation
  only once all un-sharing have been done)
- Avoid all owners to send key rotation command at the same time to the server.

### 6.3 Existing local data: `LocalUserManifest`

`LocalUserManifest` has to be modified to reflect changes in `UserManifest`:

- remove `last_processed_message` field
- change `workspace` field to use the updated `WorkspaceEntry` type

## 7 Backward compatibility & migration

As stated, we don't support APIv3 and APIv4 in the same Parsec server.

An APIv3 Parsec server can be migrate to APIv4.

> After that it will deny any APIv3 request, hence forcing all it users to upgrade
> they Parsec client at once.

Reviewer: ```suggestion
After that it will deny any APIv3 request, hence forcing all its users to upgrade
their Parsec client at once.
```

### 7.1 Server-side organization migration

This is typically for the SAAS version Parsec: some clients may want to be able

> to control when the switch from v2 to v3 should come for they organization.

Reviewer: ```suggestion
to control when the switch from v2 to v3 should come for their organizations.
```

A client can have multiple organizations

> The solution for that is to run create a new separate Parsec instance running on v3:

Reviewer: ```suggestion
The solution for that is to run a new separate Parsec instance running on v3:
```

- All existing organizations stay on the legacy Parsec v2 instance
- All new organizations are created on the new Parsec v3 instance
- When organization owner requests it, his organization can be migrated from
  legacy to the new Parsec v3 instance.

The migration goes as follow:

1) A partial database export is done from the Parsec v2 instance containing only the
  organization(s) to migrate

> 2) The partial export is a valid Parsec v2 database on it own. Hence the Parsec v2 -> v3

Reviewer: ```suggestion
2) The partial export is a valid Parsec v2 database on its own. Hence the Parsec v2 -> v3
```

>    database migration script can be run on it.

Reviewer: How the server would handle that load ?

Author: We will put a nginx proxy  on the parsec v3 server: if the user-agent is missing we will redirect the request to the parsec v2 server

> 3) Each organization of the migrated database is inserted in the Parsec v3 instance

Author: the organization must be disabled in the Parsec v2 server once migration is done, this is to avoid Parsec v2 client to access the old data (instead of parsec v3 server)

>    database. A check should be made to ensure the organization doesn't already exist.

Reviewer: That mean an additional check as to be made when creating an organization on parsec v3 to ensure that it does not already exit on  parsec v2

### 7.2 Possible compatibility strategy

A compatibility mode can be introduced and set on a per-organization basis.

> In this mode both Parsec v2 and v3 can connect to the server and use the organization,

Reviewer: ```suggestion
In this mode both Parsec v2 and v3 client can connect to the server and use the organization,
```

I prefer to be explicit

- in the server database, an organization can be marked as in APIv3 compatibility mode
- the Parsec v3 server only exposes the new APIv4
- a special Parsec v3-apiv3-compat server exists:
  - it uses the server database model of Parsec v3
  - it exposes only the APIv3
  - it only works with organization in APIv3 compatibility mode
- organization in APIv3 compatibility mode only have single encryption key that
  never changes. Hence both APIv3 maintenance system and APIv4 key-rotation systems
  are disabled.
- APIv4 `realm_update_roles` command still contains the `recipient_message` field
  along with a new `recipient_keys_bundle_access` field (that contains the `RealmKeysBundleAccess`
  structure that will be passed through `realm_get_keys_bundle`). `not_in_apiv3_compatibility_mode`
  status is returned if `recipient_message` is provided while not in compatibility mode.
- APIv4 `message_get` command is still present
- APIv4 `message_get` and `realm_update_roles`'s `recipient_message` field are not
  available outside of the compatibility mode (returns `not_in_apiv3_compatibility_mode`
  status otherwise)

> - Parsec v3 and 3-api-compat servers are behind a proxy: `/ws` route is redirected to

Reviewer 3: `v3-apiv3-compat` servers

Reviewer 3: ```suggestion
- Parsec v3 and v3-api-compat servers are behind a proxy: `/ws` route is redirected to
```

>   3-api-compat server (given it is only used by APIv3), the rest goes to regular v3 (

Reviewer 3: `v3-apiv3-compat` server ?

Reviewer 3: ```suggestion
  v3-apiv3-compat server (given it is only used by APIv3), the rest goes to regular v3 (
```

- Remove the `in_maintenance` response status
- add a `bad_key_index` response status. This error is returned if the key
  index doesn't exist for this realm, or is no longer the last one

In `vlob_read` command:

- Remove the `encryption_revision` field in the request
- add a `key_index` field (with type `Index`) in the ok response
- Remove the `bad_encryption_revision` response status
- Remove the `in_maintenance` response status

In `events_listen` command:

- Remove the `realm_maintenance_started` event
- Remove the `realm_maintenance_finished` event
- Remove the `message_received` event

### 3.2 - New command: `realm_share`

It is similar to the `realm_update_roles` with the following changes:

- Remove `in_maintenance` response status
- Remove `recipient_message` field
- `role_certificate` must be a `RealmRoleCertificate` with role != None
- Add `recipient_keys_bundle_access` field (with type `Bytes`). This contains the
  `RealmKeysBundleAccess` document encrypted for the recipient.
- add a `key_index` field (with type `IndexInt`). The server uses this to ensure
  the client is not missing a key rotation (otherwise the recipient would be missing
  some keys)

> - add a `bad_key_index` response status

Author: > notice the change for Index (from IndexInt): do we plan to have specialization for Index type ? maybe rename it to IndexUint ?

That's a typo: the type is `Index` on the json schemas (and `Index` is already a specialization, no need to go further ^^)

                    }
                ]
            },
            {
                // Returned if another certificate in the server has a creation date
                // posterior or equal to our current one.
                "status": "require_greater_timestamp",
                "fields": [
                    {
                        "name": "strictly_greater_than",
                        "type": "DateTime"
                    }
                ]
            }
        ]
    }
]
```

## 4 - Changes: Data

### 4.1 - Existing data: messages

All the messages types are removed:

- `sharing.granted`
- `sharing.reencrypted`
- `sharing.revoked`

On top of that, a special logic is added in the client to ignore those messages
types.

> **Note:**
>
> Ignoring those messages means the client may end up with invalid data (e.g. a
> user hasn't started its client in a while, he does it after the Parsec v3 migration
> and now all the workspaces it has been part of that have been re-encrypted in the
> meantime fail to load with an "decryption error" error message...).
>
> However those errors can simply be fixed by removing and re-adding the access rights

> > to the user, which is the obvious "IT, have you tried to turn it off and on again ?" way.

Author: Can you create an issue to track this ?

> > to the user, which is the obvious "IT, have you tried to turn it off and on again ?" way.

Reviewer: Done in #5767

client must make sure the `key_index` field strictly follow the value of the
previous key rotation certificate for this realm. (i.e. we reject the
certificate if there is a gap in key indexes)

`SecretKeyAlgorithm` has one possible values: `XSALSA20-POLY1305`.
`HashAlgorithm` has one possible value: `SHA256`.

> **Note:**
>
> `encryption_algorithm` & `hash_algorithm` fields are discussed in RFC-1006.
>
> In a nutshell, those fields allows for future evolution, typically to switch to
> Chacha20-Poly1305 given it is a more scrutinized algorithm with audited implementations.

### 4.4 - New certificate `RealmNameCertificate`

This certificate can only be issued by a user with OWNER role on the given realm.

`key_index` must correspond to `KeyRotationCertificate`.

```json5
{
    "label": "RealmNameCertificate",
    "type": "workspace_name_certificate",
    "other_fields": [
        {
            "name": "author",
            "type": "DeviceID",
        },
        {
            "name": "timestamp",
            "type": "DateTime"
        },
        {
            "name": "realm_id",
            "type": "VlobID"
        },
        {
            "name": "key_index",

>             "type": "Index",

Author: The correct type is `Index`

> are not one atomic operation), this is for three reasons:
>
> - It simplifies code.
> - User realm doesn't need realm name, hence this certificate would need to be optional
>   or set to a dummy value.
> - This allows legacy (i.e. realm created, but no realm name certificate) to be treated
>   just like the regular case.

### 4.5 - New document `RealmKeysBundle`

```json5
{
    "label": "RealmKeysBundle",
    "type": "realm_keys_bundle",
    "other_fields": [
        {
            // Author must be equal to the one in the `RealmRoleCertificate`
            "name": "author",
            "type": "DeviceID",
        },
        {
            // Timestamp must be equal to the one in the `RealmRoleCertificate`
            "name": "timestamp",
            "type": "DateTime"
        },
        {
            "name": "realm_id",
            "type": "VlobID"
        },
        {
            "name": "keys",
            // Keys ordered by index
            "type": "List<Bytes>"
        }
    ]
}
```

> **Note:**
>
> It could be tempting not to sign the `RealmKeysBundle`, given the keys can be
> checked against their corresponding key rotation certificate.
>

> > However this doesn't take into account that a key that doesn't match it key

Author: I corrected the it/its

It's "keys rotation certificate" without the 's (see for instance https://cloud.google.com/kms/docs/key-rotation?hl=en and https://developer.okta.com/docs/concepts/key-rotation/)

> rotation certificate must be kept in the bundle nevertheless (as removing
> it would mess the key index for the subsequent key rotations).
>
> So if the client encounter a realm keys bundle containing an invalid key it
> cannot tell if it's normal (i.e. due to a previous invalid key rotation), or
> an attack by the server (no signature means the server can craft a keys bundle
> with the right number of keys, but with all of them invalid)

### 4.6 - New document `RealmKeysBundleAccess`

```json5
{
    "label": "RealmKeysBundleAccess",
    "type": "realm_keys_bundle_access",
    "other_fields": [
        {
            "name": "keys_bundle_key",
            "type": "SecretKey"
        }
    ]
}
```

> **Note:**
>
> `RealmKeysBundleAccess` is not signed, this is because it only role it to give
> access to `RealmKeysBundle`.
>
> Hence forging a `RealmKeysBundleAccess` is possible for the server, but only
> prevent from accessing `RealmKeysBundle` which also can be done by just returning

> > invalid data in place of the actual encrypted `RealmKeysBundle` data.

Author: > If the server can send the `SecretKey` that mean it can decode the key-bundle or the SecretKey is encrypted by some means ?

user's public key is available to the server, hence the latter can send encrypted payload to the former.

So the server can generate a dummy `SecretKey`, encrypt it for alice, and pretend this is the genuine key bundle access.

Of course the server doesn't know the real `SecretKey` that encrypt the key bundle, so this trick can only lead to Alice not being able to decrypt the key bundle, which can already be achieved by the server by just sending dummy data instead of the genuine encrypted key bundle

Currently the HTTP `Server` value is `parsec`, we should change it to `parsec/3` to
simplify the [detection of an incompatible server](#7---backward-compatibility--migration)

### 5.2 - Datamodel

The main issue with data migration is handling realm currently under maintenance given
it asks the question of which key to keep:

- Keeping only the last key doesn't work as some data are not encrypted for it
- Keeping the n-1 key only makes migration complicated on client-side (the server
  no longer has concept of realm under migration, however the client has to find
  a way to know that. On top of that the migration process requires accessing
  older version of the user manifest which add complexity)
- Keeping both n-1 and n adds basically the same issues with client-side

So we go simple on this and only allow server migration if all the realms have
finished their maintenance, or if a realm doesn't contain any data (see below).

To avoid being blocked by a workspace forever in maintenance (e.g. workspace owned by
a single user that got revoked while it was re-encrypting it), we also provide a cli for
removing realm vlobs (cannot remove a realm given it is related to certificates) or

> organization form the database.

Author: this cli doesn't exist, but I think this is part of the key rotation implementation so no need for a specific issue for this for the moment

> organization form the database.

Reviewer: I've create the issue anyway (#5782), in the best case it would be merge with the implementation of the RFC

name of the realm defaults to the realm ID).

So that any realm OWNER detecting the `RealmNameCertificate` is missing should look in
it local user manifest if a workspace name is specified, and do a `realm_rename` in
that case.

This approach has the benefit of transparently handling the legacy (i.e. realm created,
but no realm name certificate).

### 6.2 - Workspaces listing

The client has to rely on both the certificates and the local user manifest to get
the list of workspaces and their corresponding name:

1) Get all the realm for which the user has a realm role certificate with currently
   active role.
2) Get the name from the `RealmNameCertificate` if available and valid.
3) Otherwise, look in the local user manifest for a workspace name entry.
4) Otherwise default to the realm ID

### 6.3 - Fetching keys-bundle

The client lazily fetch the realm keys-bundle: it does the fetch operation when a server
operation requires it (i.e. `realm_share`, `vlob_read`, `vlob_create` and `vlob_update`).

The realm keys-bundle is only kept in RAM, not on disk.

Once fetched, keys-bundle must be checked against its corresponding `RealmKeyRotationCertificate`

> for same author/timestamp/realm_id/key_index.

Reviewer: The document still state to check the quatuor `author/timestamp/realm_id/key_index` is it expected ?

Author: yes, because `KeysBundleAccess` contains those fields which are duplication of the `RealmKeyRotationCertificate` (and hence the correspondence must be checked)

Each key in the Keys-bundle must be checked against the canary in `RealmKeyRotationCertificate`
that introduced it.

At any time, `realm_share`/`vlob_read`/`vlob_create` can return `bad_key_index` if the client
consider the wrong key to be the last one. In this case the client simply forgets its realm
keys-bundle and re-do the fetch operation.

### 6.4 - Keys-bundle self-healing

A keys-bundle can be corrupted in multiple ways:

> 1. Invalid data: bad signature, bad encryption, bad serialization
> 2. Additional keys
> 3. Missing keys

Reviewer: ```suggestion
- Invalid data: bad signature, bad encryption, bad serialization
- Additional keys
- Missing keys
```

I don't think that list need to be ordered

Author: you're right (I used the number to reference the items in the next paragraph but that's no longer the case)

typically ends up as a warning logged and gets displayed in the GUI.

> **Note:**
>
> We cannot simply get the keys-bundle that originally created a given key given
> the realm might have been shared with the user later on.

### 6.5 - Automatic key rotation

For automatic key rotation, a new "key rotation manager" is to be introduced. Its function
is very simple:

1. Wait for new `UserRevokedCertificate`, `RealmRoleCertificate` and
   `RealmKeyRotationCertificate` certificates
2. For `UserRevokedCertificate`, call `realm_unshare` for each realm we are owner (and
   the revoked user is part of)
3. For `RealmRoleCertificate`, if the role is None and we are owner of the realm:
    1. Wait a random time between 30 and 60 seconds (and reset the wait if we were
       already waiting for this realm)
    2. Call `realm_rotate_key`
4. For `RealmKeyRotationCertificate` cancel the wait from step 3.1 if any

The wait at step 3.1 is here for two reasons:

- Give priority to the author of the certificates to do the key rotation. This is
  typically needed if he is doing un-sharing batch (and hence will do the key rotation

>   only once all un-sharing have been done)

Reviewer: If the owner is in step `3.1` (waiting ~45s) If the owner receive another `RealmRoleCertifcate` what should append ?

1. The step `3.1` is reset and a new time period is awaited.
2. The step `3.1` continue (meaning that an user as at most ~30s to do a batch un-share). 

The document isn't clear about which behavior to use

Author: The correct answer is
> The step 3.1 is reset and a new time period is awaited.

That's what is explained by ` and reset the wait if we were  already waiting for this realm)` in 3.1.

I've reworded it a bit to make it clearer

> - Avoid all owners to send key rotation command at the same time to the server.

Reviewer: Given 2 owner that send a key-rotation command at the "same" time (we are unlucky, or not :thinking:), what would append ?

Is it first arrived, first served ?

Author: > Is it first arrived, first served ?

yes, the server checks the key rotation index against the last one in database and returns a `bad_key_index` status

### 6.6 - Existing local data: `LocalUserManifest`

`LocalUserManifest` has to be modified to reflect changes in `UserManifest`:

- remove `last_processed_message` field
- change `workspace` field to use a updated `WorkspaceEntry` type

> ```json

Reviewer: ```suggestion
```jsonc
```

or

```suggestion
```json5
```

      // name of the workspace on creation.
      // This only needed in the brief moment before the `RealmNameCertificate` is uploaded
      // (which can take some time if e.g. the workspace is created while the client is offline)
      "name": "name",
      "type": "RequiredOption<EntryName>"
    }
    // Legacy ignored fields
    // - `key`
    // - `encryption_revision`
    // - `encrypted_on`
    // - `role_cache_timestamp`
    // - `role`
  ]
}
```

### 6.7 - Add HTTP `User-Agent` header

Currently the client send no `User-Agent` header to the server.

Instead we should set it `Parsec-client/<version>`. (e.g. `Parsec-Client/3.0.1`).

This allows us to use reverse proxy in front of the SAAS server to
[deal with Parsec v2 vs Parsec v3 client](#71---server-side-organization-migration)

## 7 - Backward compatibility & migration

As stated, we don't support APIv3 and APIv4 in the same Parsec server.

An APIv3 Parsec server can be migrate to APIv4.

After that it will deny any APIv3 request, hence forcing all it users to upgrade
their Parsec client at once.

### 7.1 - Server-side organization migration

> This is typically for the SAAS version Parsec: some clients may want to be able
> to control when the switch from v2 to v3 should come for their organization.

Reviewer: ```suggestion
This is typically for the SAAS version Parsec: 

Some clients may want to be able to control when the switch from v2 to v3 should come for their organization.
```

Author: I don't think this is needed: the sentence is already small enough.
